HV-1463 Add test for using @ScriptAssert under OSGi

marko-bekhta opened the pull request:
- https://hibernate.atlassian.net/browse/HV-1463

I've added a simple test for `ScriptAssert`. The problem described in the issue is caused by how OSGi works with script engines. What I found while reading through similar issues - is that engines are not picked up correctly by OSGi and as a result an exception from `javax.script` is thrown. So to fix

> java.lang.ClassNotFoundException: javax.script.ScriptException ....

I just added one more import for `javax.script.*`. But the problem remains - engines are not picked up. I'm not very familiar with the OSGi and from what I found on OSGi+JSR-233 topic is that people recommend to manually create engines rather than request them like this `new ScriptEngineManager().getEngineByName( languageName )`. Which I don't think what we should do...

So maybe you have an  idea how this should be fixed correctly? If needed current exception is next:
```
javax.script.ScriptException: No JSR 223 script engine found for language "groovy".

	at org.hibernate.validator.internal.util.scriptengine.ScriptEvaluatorFactory.createNewScriptEvaluator(ScriptEvaluatorFactory.java:90)
	at org.hibernate.validator.internal.util.scriptengine.ScriptEvaluatorFactory.getScriptEvaluatorByLanguageName(ScriptEvaluatorFactory.java:70)
	at org.hibernate.validator.internal.constraintvalidators.hv.ScriptAssertContext.getScriptEvaluator(ScriptAssertContext.java:60)
	at org.hibernate.validator.internal.constraintvalidators.hv.ScriptAssertContext.<init>(ScriptAssertContext.java:34)
	at org.hibernate.validator.internal.constraintvalidators.hv.ScriptAssertValidator.initialize(ScriptAssertValidator.java:43)
	at org.hibernate.validator.internal.constraintvalidators.hv.ScriptAssertValidator.initialize(ScriptAssertValidator.java:28)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintValidatorManager.initializeValidator(ConstraintValidatorManager.java:225)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintValidatorManager.createAndInitializeValidator(ConstraintValidatorManager.java:151)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintValidatorManager.getInitializedValidator(ConstraintValidatorManager.java:109)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintTree.getInitializedConstraintValidator(ConstraintTree.java:174)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintTree.validateConstraints(ConstraintTree.java:111)
	at org.hibernate.validator.internal.engine.constraintvalidation.ConstraintTree.validateConstraints(ConstraintTree.java:82)
	at org.hibernate.validator.internal.metadata.core.MetaConstraint.doValidateConstraint(MetaConstraint.java:117)
	at org.hibernate.validator.internal.metadata.core.MetaConstraint.validateConstraint(MetaConstraint.java:110)
	at org.hibernate.validator.internal.engine.ValidatorImpl.validateMetaConstraint(ValidatorImpl.java:552)
	at org.hibernate.validator.internal.engine.ValidatorImpl.validateConstraintsForSingleDefaultGroupElement(ValidatorImpl.java:510)
	at org.hibernate.validator.internal.engine.ValidatorImpl.validateConstraintsForDefaultGroup(ValidatorImpl.java:479)
	at org.hibernate.validator.internal.engine.ValidatorImpl.validateConstraintsForCurrentGroup(ValidatorImpl.java:444)
	at org.hibernate.validator.internal.engine.ValidatorImpl.validateInContext(ValidatorImpl.java:394)
	at org.hibernate.validator.internal.engine.ValidatorImpl.validate(ValidatorImpl.java:183)
	at org.hibernate.validator.osgi.integrationtest.OsgiIntegrationTest.canUseScripAssertConstraint(OsgiIntegrationTest.java:269)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.ops4j.pax.exam.invoker.junit.internal.ContainerTestRunner.runLeafWithRetry(ContainerTestRunner.java:97)
	at org.ops4j.pax.exam.invoker.junit.internal.ContainerTestRunner.runChildWithRetry(ContainerTestRunner.java:84)
	at org.ops4j.pax.exam.invoker.junit.internal.ContainerTestRunner.runChild(ContainerTestRunner.java:75)
	at org.ops4j.pax.exam.invoker.junit.internal.ContainerTestRunner.runChild(ContainerTestRunner.java:44)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:115)
	at org.ops4j.pax.exam.invoker.junit.internal.JUnitProbeInvoker.invokeViaJUnit(JUnitProbeInvoker.java:124)
	at org.ops4j.pax.exam.invoker.junit.internal.JUnitProbeInvoker.findAndInvoke(JUnitProbeInvoker.java:97)
	at org.ops4j.pax.exam.invoker.junit.internal.JUnitProbeInvoker.call(JUnitProbeInvoker.java:73)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.ops4j.pax.exam.rbc.internal.RemoteBundleContextImpl.remoteCall(RemoteBundleContextImpl.java:83)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at sun.rmi.server.UnicastServerRef.dispatch(UnicastServerRef.java:324)
	at sun.rmi.transport.Transport$1.run(Transport.java:200)
	at sun.rmi.transport.Transport$1.run(Transport.java:197)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.Transport.serviceCall(Transport.java:196)
	at sun.rmi.transport.tcp.TCPTransport.handleMessages(TCPTransport.java:568)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run0(TCPTransport.java:826)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.lambda$run$0(TCPTransport.java:683)
	at java.security.AccessController.doPrivileged(Native Method)
	at sun.rmi.transport.tcp.TCPTransport$ConnectionHandler.run(TCPTransport.java:682)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

gsmet commented:
Getting Groovy to work is a nice second step but can you try to get it to work with Nashorn first? It's included in the JDK so it should be easier. AFAICS, they fixed issues with it in Karaf 4 so it's supposed to work.

As for getting Groovy to work, the ScriptManager is using the TCCL by default so we probably need the same sort of magic we do for the EL factory. See https://github.com/hibernate/hibernate-validator/blob/master/engine/src/main/java/org/hibernate/validator/messageinterpolation/ResourceBundleMessageInterpolator.java for how we initialize the EL trying different CL until we find one working.

We should also take into account the `externalClassLoader`. See what we do here: https://github.com/hibernate/hibernate-validator/blob/ff2ff96b1f30f198c37109f3b7825bc9ff8c4779/engine/src/main/java/org/hibernate/validator/internal/engine/ConfigurationImpl.java#L546 . The issue is that you will probably need to pass it to the `HibernateConstraintValidatorContext` through the `ValidationContext`. That shouldn't be too hard but that would put some pressure on the `ScriptEvalutor`map as we wouldn't be able to interrogate it only once at the validator initialization.

@gunnarmorling what's your thoughts on this?

marko-bekhta commented:
@gsmet thanks for the info! I did find one more thing, which I'm half-way finished implementing. The idea is to try and get ScriptEngine the way it is right now, and if nothing is found then try to search for it through OSGi bundles. So I think it's similar to what you were saying about the EL. The only difference would be that we will not be able to use something like `GroovyScriptEngine.class.getClassLoader` as we don't know which scripting engine is requested by the user ...

I think I'll have something working later today.

marko-bekhta commented:
I used a couple of classes from https://svn.apache.org/repos/asf/felix/trunk/mishell/src/main/java/org/apache/felix/mishell/  as they seem to do what we need, and I thought "why reinvent the wheel"? I added the classes and in the next commit I slightly modified `findFactoryCandidates` method as it was not handling files correctly (adding comments and empty lines as candidates) + reformatted to match HV formatting. 

Right now it's not very clean but if we agree that using these is OK - I'll clean it all up (move the dependencies to dependency management section, remove those retries of `getEngineByName` in `ScriptEvaluatorFactory` and hide it all behind selection of correct `ScriptEngineManager`; also those `getEngineBy***` methods should probably be using something like `run( SetContextClassLoader.action( ResourceBundleMessageInterpolator.class.getClassLoader() ) );` as well ?)

gsmet commented:
@marko-bekhta I'm not very convinced about this approach even if it solves the issue (and kudos for finding a way to do it).

Even if we clean it up, it's still quite a hack so I wouldn't like to have it in plain HV.

Another approach might be to provide a way to inject a `ScriptEvaluatorFactory` (not exactly in this form, we would need to make an API out of it) from the user space to the `HibernateConstraintValidatorContext`.

That way, people would have the ability to extend it the way they want (and you could clean up your impl and provide a jar for it).

WDYT? Does it make sense to you?

gunnarmorling commented:
On a related matter, I've just filed [HV-1469](https://hibernate.atlassian.net/browse/HV-1469) for managing the lifecycle of SEF in the scope of the validator factory instead of the static singleton it is currently.

gunnarmorling commented:
Btw. `ScriptEngineManager` can be instantiated with a dedicated classloader, so we could try to pass the `externalClassLoader` for that. I'm doubting it helps much for the OSGi issue, but it likely is a good idea nevertheless, as we should consult that CL for all resources we try to look up.

marko-bekhta commented:
hmm ... You have good points. 
It got me thinking ... I kind of like the idea of passing the `ScriptEngineFactory` as in this case we don't need lookups at all, but the users should preconfigure HV anytime they want to use `@ScriptAssert`.
If we pass the `ScriptEvaluatorFactory` then we can use the current implementation by default (adding an `externalClassLoader` to it as Gunnar mentioned), and if somebody needs an OSGi version we can give them a "cleaned up version" from this PR... 

In case of passing engine factories I think it'll be enough to have an interface like:
```java
public interface HVScriptEngineFactory {
    public String getLanguageName();
    public  ScriptEngine getScriptEngine(); // or maybe return ScriptEvaluator right away rather then ScriptEngine
}
```
Which is a subset of methods from `ScriptEngineFactory`. As we just need an `ScriptEngine` itself, and a language name - so we can identify and cache them by name. 

If the idea with `ScriptEngineFactory`s is not a good solution :) I'll proceed with `ScriptEvaluatorFactory` then. And in this case I think the only method that we need to expose is `ScriptEvaluatorFactory#createNewScriptEvaluator`.

gsmet commented:
@marko-bekhta not sure I follow you. The point would be to pass an API equivalent of `ScriptEvaluatorFactory` from the `ValidationFactory` and you would be able to override it at bootstrap. So you wouldn't have to preconfigure anything as it would use the default one if not overridden.

Maybe I missed something?

marko-bekhta commented:
@gsmet right that's one option and in case of overriding it looks that the only method needed is similar to `ScriptEvaluatorFactory#createNewScriptEvaluator`.

And as another option we make `ScriptEvaluatorFactory` just a "storage" of `HVScriptEngineFactory` something like:
```java
class ScriptEvaluatorFactory {
   Map<String, HVScriptEngineFactory> mapOfEngineFactories;
   //...

  public ScriptEvaluator getScriptEvaluatrorForLanguage(String name){
      return mapOfEngineFactories.get(name). getScriptEngine();
  }
}
```
And these `HVScriptEngineFactory` will be passed by user at bootstrap:
```java
Validator validator = Validation.byProvider( HibernateValidator.class )
        .configure()
        .scripEngineFactories(
            Arrays.asList( new HVGroovyScriptEngineFactory(), new HVRubyScriptEngineFactory() )
        ).buildValidatorFactory()
        .getValidator();
```

gsmet commented:
It kinda defeats the whole point of the ScriptEngineManager dynamically finding the script implementations.

The advantage of my proposition is that people not running OSGi wouldn't need any configuration as the default implementation would rely on the JDK `ScriptEngineManager` (probably passing the externalClassLoader as the class loader if defined).

And if you want something like your proposal, you could still have something like:
```java
Validator validator = Validation.byProvider( HibernateValidator.class )
        .configure()
        .scripEngineFactory(
            new DeclarativeScriptEngineFactory( new HVGroovyScriptEngineFactory(), new HVRubyScriptEngineFactory() )
        ).buildValidatorFactory()
        .getValidator();
```
(either provided by HV or not depending on how clean we can make it)

marko-bekhta commented:
I've changed the approach as it was suggested. It still needs some things here and there (missing docs, OSGi factory which takes bundle (similar to what was in the original version), finishing implementation in xml (need to confirm if it's ok to "extend" the schema and add `ScriptEvaluatorFactory` to it, I assume it'll be HV specific schema in this case, right?).

Also I'm not very happy about the fact that it seems that we need to add package imports for each scripting language we want to be used (that list should be maintained and what about some custom user implementations, which we know nothing about ?)

But in general - that's what we'd like to have, right ?

gsmet commented on engine/src/main/xsd/hibernate-validation-configuration-2.0.xsd:
```diff
@@ -0,0 +1,26 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This is a bold move :).

We'll have to wait for Gunnar for that (he's on vacation). I'm not sure we want to extend the schema for Hibernate specific features.

Wondering if it wouldn't be better to use a `<property>`, especially considering it's such a niche feature and it fits well.

marko-bekhta replied:
:) yeah ... I wasn't sure about this as well - if it's a right way to do it. And also about that comment on `validation-custom-config.xml` it's commented out as the test failed and I didn't have time to figure out what happened...

marko-bekhta replied:
@gsmet I've removed the schema, so hope that Gunnar won't see that :). I changed the code to use the property and if the name of the property is OK, I think I could add some paragraph to documentation on how to set the `ScriptEvaluatorFactory` why it's needed and the ways it can be done.

marko-bekhta commented on pom.xml:
```diff
@@ -167,7 +167,7 @@
         <puppycrawl.checkstyle.version>7.6</puppycrawl.checkstyle.version>
 
-        <groovy.version>2.4.8</groovy.version>
+        <groovy.version>2.4.12</groovy.version>
```
OK we can do that. I just saw that there's a newer version and updated it as I was using that dependency. I'll create a ticket for it.

marko-bekhta commented on pom.xml:
```diff
@@ -292,6 +292,11 @@
             <dependency>
+                <groupId>org.osgi</groupId>
+                <artifactId>org.osgi.core</artifactId>
+                <version>6.0.0</version>
```
OK :)  I just wasn't 100% sure if we'd keep it or not because of that OSGiFactory thing. So I left it "in a bad sate" so we'd see it :)

gsmet replied:
Yes, tbh, I'm not sure either. But let's make it clean, it's not as if it was a lot of work.

marko-bekhta commented on engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorFactoryImpl.java:
```diff
@@ -150,6 +156,7 @@ public ValidatorFactoryImpl(ConfigurationState configurationState) {
 		this.traversableResolver = configurationState.getTraversableResolver();
 		this.parameterNameProvider = new ExecutableParameterNameProvider( configurationState.getParameterNameProvider() );
 		this.clockProvider = configurationState.getClockProvider();
+		this.scriptEvaluatorFactory = ( (ConfigurationImpl) configurationState ).getScriptEvaluatorFactory();
```
@gsmet I didn't extract the getter to the in interface. Extract it and then change the casting ?

gsmet replied:
+1

marko-bekhta commented on engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/ScriptAssertContext.java:
```diff
@@ -27,25 +29,26 @@
 	}
 
-	public boolean evaluateScriptAssertExpression(Map<String, Object> bindings) {
+	public boolean evaluateScriptAssertExpression(Map<String, Object> bindings, ConstraintValidatorContext context) {
```
Right ... But it's either here or via the constructor right? And in case of constructor we'd need to get it in the `initailize` method:

```java
public void initialize(ParameterScriptAssert constraintAnnotation) {
   // get the Context somehow
}
```

Or did you have something different on your mind?

gsmet replied:
No it was more about building what's needed in `ParameterScriptAssertValidator` and pass the `ScriptEvaluator` directly.

About the `initialize()`, that's what I posted in another message, it's too bad we don't have a `Context` object here, it would have been useful.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/ScriptAssertContext.java:
```diff
@@ -54,14 +58,16 @@ public boolean evaluateScriptAssertExpression(Map<String, Object> bindings) {
-		}
-		catch (ScriptException e) {
-			throw log.getCreationOfScriptExecutorFailedException( languageName, e );
+	private synchronized ScriptEvaluator getScriptEvaluator( ConstraintValidatorContext context) {
```
So if we move that to the Validator, please don't use a plain `synchronized` method but take a look at http://www.javamex.com/tutorials/double_checked_locking_fixing.shtml , especially the last paragraph: Use DCL plus volatile (except the synchronized should be on (this) not on the class).

gsmet commented on engine/src/main/java/org/hibernate/validator/HibernateValidatorConfiguration.java:
```diff
@@ -62,6 +63,16 @@
+	 *
+	 * @since 6.1
+	 */
+	String SCRIPT_EVALUATOR_CLASSNAME = "hibernate.validator.script_evaluator_factory_classname";
```
+1 but you can remove the _classname suffix, it's useless.

marko-bekhta commented:
@gsmet Thanks! I'm writing the doc section about how this Script Factory can be configured and I think I'll finish it all till the end of today (doc + some clean ups that left).

gsmet commented:
OK, nice! No pressure!

marko-bekhta commented:
@gsmet I've forgot to push the changes :) 

I've added documentation section and also I've changed how the `ScriptAssertContext` is created. I've used the double checking as you've suggested and also removed the `HibernateConstraintValidatorContext` parameter from the evaluation method and created an abstract validator for both `Script` constraints and placed the code for evaluator creation there.  I've also added that `externalClassLoader` to the default factory.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/util/scriptengine/DefaultLookupScriptEvaluatorFactory.java:
```diff
@@ -29,11 +26,6 @@
  */
 public class DefaultLookupScriptEvaluatorFactory extends AbstractCacheableScriptEvaluatorFactory {
 
-	/**
```
@marko-bekhta ah have a misunderstanding here. I was asking to remove the `SoftReference` but keep the singleton instance, just not with a `SoftReference` :).

marko-bekhta replied:
right ... It looks like that that's how I addressed this ticket - https://hibernate.atlassian.net/browse/HV-1469 
And then when I've added that part with the external classloader we cannot cache the instance unless it's just one external loader (and it's passed to `getInstance()`), right?

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,100 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+[[section-script-evaluator-factory]]
+==== Configuring script evaluator factory (`ScriptEvaluatorFactory`)
+
+For constrains like `@ScriptAssert` and `@ParameterScriptAssert` it might be useful to configure
```
constrain**t**s
add a comma before "it might", it's more readable.

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,100 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+==== Configuring script evaluator factory (`ScriptEvaluatorFactory`)
+
+For constrains like `@ScriptAssert` and `@ParameterScriptAssert` it might be useful to configure
+how scripts are evaluated, and which scripting engines are used. This can be done by setting
```
It's more about configuring how the script engines are built?

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,100 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+For constrains like `@ScriptAssert` and `@ParameterScriptAssert` it might be useful to configure
+how scripts are evaluated, and which scripting engines are used. This can be done by setting
+a custom implementation of `ScriptEvaluatorFactory`. In particular this is important for
+OSGi environments (or similar to it) where user might face issues with dynamic class loading
```
s/dynamic/modular/ ?

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,100 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+
+===== Programmatic configuration
+There are two ways to change the default `ScriptEvaluatorFactory` programmatically.
+First one is to use `hibernate.validator.script_evaluator_factory` property as described
```
I don't think I would advertise the property configuration in the programmatic API. I would only use it for the XML configuration.

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,100 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+
+Another option is to pass an instance of `ScriptEvaluatorFactory` when building `ValidatorFactory`.
+This gives more flexibility in configuration of the `ScriptEvaluatorFactory`. <<example-script-evaluator-factory-programmatically>>
+shows how this can be done.
```
Should be reworked because of the remark above. But just as I think this part will be kept: "when building **the** `ValidatorFactory`"

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,100 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+
+===== XML configuration
+To specify `ScriptEvaluatorFactory` via XML the same `hibernate.validator.script_evaluator_factory`
+property can be used.
```
This should be changed accordingly.

+ To specify **the** `ScriptEvaluatorFactory`

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,100 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+====
+
+In case of using property to configure a factory (either in XML configuration, or by adding it programmatically)
+user is limited to use factories that has constructors without parameters.
```
I would simply have this:
In this case, the specified `ScriptEvaluatorFactory` must have a no-arg constructor.

(this is the wording we use in the spec so let's be consistent)

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,100 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+user is limited to use factories that has constructors without parameters.
+
+===== Predefined `ScriptEvaluatorFactory` implementations
+Hibernate Validator provides multiple implementations of `ScriptEvaluatorFactory` that user
```
OK, I have a couple more comments for the rest of the wording but it's probably easier if I fix it on merging.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/AbsctractScriptAssertValidator.java:
```diff
@@ -0,0 +1,52 @@
+/**
+ * @author Marko Bekhta
+ */
+public abstract class AbsctractScriptAssertValidator<A extends Annotation, T> implements ConstraintValidator<A, T> {
```
**Abstract**

marko-bekhta commented:
@gsmet thanks for reviewing. I addressed your suggestions and switched places of XML and programmatic configuration so the part about programmatic one having more flexibility doesn't look out of place.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/HibernateValidatorBootstrapConfiguration.java:
```diff
@@ -0,0 +1,29 @@
+	 *
+	 * @since 6.1
+	 */
+	default String getScriptEvaluatorFactoryClassName() {
```
I don't think we should add this method. One can get the configured value via `BootstrapConfiguration#getProperties()`. And `BootstrapConfiguration` is not meant to expose the implementation-specific class names, instead there should be `HibernateValidatorConfiguration#getDefaultScriptEvaluatorFactory()` which returns an _instance_.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/HibernateValidatorConfiguration.java:
```diff
@@ -62,6 +63,16 @@
 
 	/**
+	 * Property for configuring script evaluator factory, allowing to set up which factory will be used to create
+	 * {@link org.hibernate.validator.cfg.scriptengine.ScriptEvaluator} for evaluation of script expressions in
```
Let's use unqualified names (via imports) in JavaDocs for the sake of readability of rendered docs.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/HibernateValidatorConfiguration.java:
```diff
@@ -210,4 +221,26 @@
+	 *
+	 * @since 6.1
+	 */
+	HibernateValidatorConfiguration scriptEngineFactory(ScriptEvaluatorFactory scripEngineFactory);
```
s/scriptEngineFactory/scriptEvaluatorFactory/

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/HibernateValidatorFactory.java:
```diff
@@ -18,6 +20,16 @@
 public interface HibernateValidatorFactory extends ValidatorFactory {
 
 	/**
+	 * Returns a factory which is used to create {@link org.hibernate.validator.cfg.scriptengine.ScriptEvaluator}s for
```
We can use non-qualified names.

gunnarmorling replied:
Why is btw. that this method is exposed here. What's is use case?

marko-bekhta replied:
It was done by analogy with the `ClockProvider` (for which there's `ValidatorFactory#getClockProvider()` ), so that I don't miss any places where this new ScriptEvaluator is needed. But looking at this method - it doesn't seems to be necessary. I'll remove it.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/cfg/scriptengine/ScriptEvaluator.java:
```diff
@@ -0,0 +1,35 @@
+	 *
+	 * @throws ScriptException in case of any errors during script evaluation
+	 */
+	Object evaluate(String script, Map<String, Object> bindings) throws ScriptException;
```
I didn't have an abstraction as such in mind, but I think it's a good idea. But if we want to go for it, the exception in the signature shouldn't be from `javax.script` ;)

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/cfg/scriptengine/impl/AbstractCacheableScriptEvaluatorFactory.java:
```diff
@@ -0,0 +1,69 @@
+ * @author Kevin Pollet &lt;kevin.pollet@serli.com&gt; (C) 2011 SERLI
+ * @author Marko Bekhta
+ */
+public abstract class AbstractCacheableScriptEvaluatorFactory implements ScriptEvaluatorFactory {
```
I don't think we should provide any user-facing implementations (assuming that this is what you intended to do with that `impl` package). We can add a sketch for OSGi to the docs, but I don't feel it's something I'd like to maintain as part of the library proper.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/cfg/scriptengine/impl/AbstractCacheableScriptEvaluatorFactory.java:
```diff
@@ -0,0 +1,69 @@
+	 */
+	@Override
+	public ScriptEvaluator getScriptEvaluatorByLanguageName(String languageName) {
+		if ( !scriptExecutorCache.containsKey( languageName ) ) {
```
We could use calculateIfAbsent() now that we are on Java 8.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/cfg/scriptengine/impl/OSGiScriptEvaluatorFactory.java:
```diff
@@ -0,0 +1,111 @@
+		return findFactoryCandidates( context ).stream()
+				.map( className -> {
+					try {
+						return new ScriptEngineManager( Class.forName( className ).getClassLoader() );
```
I don't think using `Class#forName()` is a good idea in OSGi. I'd expect people will rather plug-in just a very simple implementation of `ScriptEvaluatorFactory` which just returns an implementation they statically passed via the constructor.

marko-bekhta replied:
this factory is based on [1] and [2], and how script engines are handled there. But if as you said earlier - we will not support factory implementations for users to choose from - we probably wouldn't need this `OSGiScriptEvaluatorFactory` anyway :)

[1]https://svn.apache.org/repos/asf/felix/trunk/mishell/src/main/java/org/apache/felix/mishell/OSGiScriptEngineManager.java
[2]https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/osgi/impl/OSGiScriptEngineManager.java

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/constraintvalidation/HibernateConstraintValidatorContext.java:
```diff
@@ -6,8 +6,11 @@
+import javax.script.ScriptException;
 import javax.validation.ConstraintValidatorContext;
 
+import org.hibernate.validator.cfg.scriptengine.ScriptEvaluator;
```
That package seems a bit odd, how about `org.hibernate.validator.script(ing)`?

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/constraintvalidation/HibernateConstraintValidatorContext.java:
```diff
@@ -106,4 +109,19 @@
+	 *
+	 * @since 6.1
+	 */
+	ScriptEvaluator getScriptEvaluatorForLanguage(String languageName) throws ScriptException;
```
Nice, that's much better than before.

gunnarmorling commented on engine/src/test/java/org/hibernate/validator/test/internal/constraintvalidators/hv/ScriptAssertValidatorTest.java:
```diff
@@ -93,34 +105,52 @@ public void emptyAliasRaisesException() throws Exception {
 		getInitializedValidator( "lang", "script", "" );
 	}
 
-	@Test(expectedExceptions = ConstraintDeclarationException.class)
```
What's the reason to change that? I find that much more expressive personally.

marko-bekhta replied:
I do like the `expectedExceptions` more compared to this change. But as
now that we only try to create a `ScriptEvaluator` when `isValid` is called. We are not able to check for things like "is there an evaluator for specified language?" during initialization. As a result the exception (`ConstraintDeclarationException`) is only thrown from isValid method and it is wrapped with `ValidationException` ...

gunnarmorling replied:
Ok.

marko-bekhta commented:
So I finally got back :)
I think I've addressed all the comments. 
I've also added a test for SpEL (see `CustomScriptEvaluatorFactoryForSpringELTest`) as @gunnarmorling suggested . It seems to be working with the current interfaces for factory and evaluator, but personally I'd go for something like 
```java
public interface ScriptEvaluator {

	/**
	 * Evaluates a {@code script} expression and returns a result of this evaluation.
	 *
	 * @param script A script to evaluate
   * @param alias The alias for the object used in script
   * @param object The object against which to evaluate the script  
	 *
	 * @return The result of script evaluation
	 *
	 * @throws ScriptEvaluationException in case of any errors during script evaluation
	 */
	Object evaluate(String script, String alias, Object object) throws ScriptEvaluationException;

}
```
WDYT? Should we go for such change or keep it as is?

I've also left the documentation part unchanged for now. I wanted to confirm first  if we should remove those factory implementations completely or just move them from engine to examples and show them in the documentation ?

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,86 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+==== Configuring script evaluator factory (`ScriptEvaluatorFactory`)
+
+For constraints like `@ScriptAssert` and `@ParameterScriptAssert`, it might be useful to configure
+how script evaluators (`ScriptEvaluator`) are build, and which scripting engines are used to build
```
s/build/built/

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,86 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+For constraints like `@ScriptAssert` and `@ParameterScriptAssert`, it might be useful to configure
+how script evaluators (`ScriptEvaluator`) are build, and which scripting engines are used to build
+them. This can be done by setting a custom implementation of `ScriptEvaluatorFactory`. In
+particular this is important for OSGi environments (or similar to it) where user might face issues
```
"this is important for modular environments (e.g. OSGi)" instead of "this is important for OSGi environments (or similar to it)"

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,86 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+In this case, the specified `ScriptEvaluatorFactory` must have a no-arg constructor.
+
+===== Programmatic configuration
+To do the change programmatically an instance of `ScriptEvaluatorFactory` should be passed when building the `ValidatorFactory`.
```
"To do the change programmatically" -> "To configure it programmatically, "

gsmet commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -370,6 +370,86 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/Bootstrapp
+====
+
+===== Predefined `ScriptEvaluatorFactory` implementations
+Hibernate Validator provides multiple implementations of `ScriptEvaluatorFactory` that user
```
OK, I'll let you change that, considering we won't include both by default.

gsmet commented on documentation/src/test/resources/org/hibernate/validator/referenceguide/chapter09/script-evaluator-factory-validation.xml:
```diff
@@ -0,0 +1,10 @@
+        version="2.0">
+
+    <property name="hibernate.validator.script_evaluator_factory">org.hibernate.validator.referenceguide.chapter09.CustomScriptEvaluatorFactory</property>
+
```
Add a new line at the end of file.

gsmet commented on engine/src/main/java/org/hibernate/validator/constraintvalidation/HibernateConstraintValidatorContext.java:
```diff
@@ -106,4 +110,19 @@
+	 *
+	 * @since 6.1
+	 */
+	ScriptEvaluator getScriptEvaluatorForLanguage(String languageName) throws ScriptException;
```
Mmmmh, should it be a `ScriptException`? I thought we were trying to get away from `javax.script` specific exceptions?

gsmet commented on engine/src/main/java/org/hibernate/validator/scripting/impl/OSGiScriptEvaluatorFactory.java:
```diff
@@ -0,0 +1,110 @@
+					.collect( Collectors.toList() );
+		}
+		catch (IOException e) {
+			throw new ValidationException( "Wasn't able to read a ScripFactory candidate resource file", e );
```
s/ScripFactory/ScriptEvaluatorFactory/

gsmet commented on engine/src/main/java/org/hibernate/validator/scripting/impl/AbstractCacheableScriptEvaluatorFactory.java:
```diff
@@ -0,0 +1,66 @@
+			return createNewScriptEvaluator( languageName );
+		}
+		catch (ScriptEvaluationException e) {
+			throw new ConstraintDeclarationException( "Wasn't able to find a script evaluator for '" + languageName + "' language.", e );
```
It would probably be better to move the exception creation in the logger for the "core" script evaluator factory.

gsmet commented on engine/src/main/java/org/hibernate/validator/scripting/ScriptEvaluatorFactory.java:
```diff
@@ -0,0 +1,35 @@
+public interface ScriptEvaluatorFactory {
+
+	/**
+	 * Retrieves a script executor {@link ScriptEvaluator} for the given language.
```
s/executor/evaluator/

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/xml/ValidationBootstrapParameters.java:
```diff
@@ -243,6 +258,22 @@ private void setClockProvider(String clockProviderFqcn, ClassLoader externalClas
+				scriptEvaluatorFactory = run( NewInstance.action( clazz, "script evaluator factory class" ) );
+				log.usingScriptEvaluatorFactory( clazz );
+			}
+			catch (ValidationException e) {
```
I think we'd better catch any `Exception` here.

gsmet commented on engine/src/main/java/org/hibernate/validator/internal/util/scriptengine/ScriptEvaluatorImpl.java:
```diff
@@ -52,8 +57,13 @@ public Object evaluate(String script, Map<String, Object> bindings) throws Scrip
+		try {
+			return engine.eval( script, new SimpleBindings( bindings ) );
+		}
+		catch (ScriptException e) {
```
Let's catch any exception here to be on the safe side.

marko-bekhta commented:
@gsmet no worries :) I kind of like these tasks "hmm it's just a 10 minutes probably to add a test" which turn out into quite some changes. I'll update the doc and let you know when it's ready for the next review :).

marko-bekhta commented:
Hi @gsmet 
I've done some more cleanup around exceptions and updated the documentation. As it was suggested I moved out factory impls out of engine. But what I don't like about it is that there are now two copies of OSGi and MultiClassloader factories as one pair is used in documentation and another - in osgi module to run actual tests in OSGi env....

gsmet commented:
@marko-bekhta I rebased and pushed a few other commits. I think we're in pretty good shape now. Interested in your feedback on my commits!

@gunnarmorling there are 2 things I wanted to check with you:
* should we commit this in the 6.0 branch or do we keep it for 6.1. I know you were a bit reluctant about creating a 6.1. The current patch does not break any API and fixes an issue with the ScriptEvaluatorFactory currently being a singleton
* I made a few implementations visible in the API (namely `AbstractCachingScriptEvaluatorFactory` and `ScriptEngineScriptEvaluator`): we want them to be accessible so that users can base their own implementations on it. The 2 implementations Marko created for OSGi now use both classes (Marko initially didn't use them but I don't think it's such a good idea).

marko-bekhta commented on documentation/src/main/asciidoc/ch09.asciidoc:
```diff
@@ -445,21 +452,22 @@ include::{sourcedir}/org/hibernate/validator/referenceguide/chapter09/OSGiScript
+As already mentioned, you can also use script engines that are not based on https://www.jcp.org/en/jsr/detail?id=223[JSR 223].
+
+For instance, to use the https://docs.spring.io/spring/docs/current/spring-framework-reference/core.html#expressions[Spring Expression Language]),
+you can define a `SpringELEvaluatorFactory` as:
```
Probably `SpringELEvaluatorFactory` -> `SpringELScriptEvaluatorFactory` as the name changed

gsmet replied:
Good catch, fixed.

marko-bekhta commented on engine/pom.xml:
```diff
@@ -229,9 +224,7 @@
                             org.jsoup.*;version="[1.5.2,2.0.0)";resolution:=optional,
                             javax.money;version="[1.0.0,2.0.0)";resolution:=optional,
-                            com.thoughtworks.paranamer.*;version="[2.5.5,3.0.0)";resolution:=optional,
-                            org.codehaus.groovy.jsr223;version="[2.4.0,3.0.0)";resolution:=optional,
```
hmm so now that those ScriptEvaluatorFactories are not inside of HV engine there's no need to import scripting packages, that's really nice!

gunnarmorling commented on documentation/src/test/java/org/hibernate/validator/referenceguide/chapter09/CustomScriptEvaluatorFactoryForSpringELTest.java:
```diff
@@ -0,0 +1,42 @@
+	}
+
+	//tag::include[]
+	@ScriptAssert(script = "value > 0", lang = "spring")
```
Very nice to see this example for using SpEL! It doesn't seem to work, though, to use the "_this" alias (see `@ScriptAssert#alias()`). Unlike JSR 223, SpEL resolves the properties against a root object, I'm wondering whether we could resolve references against `_this` (or whatever is chosen as name), too, against that root object. It's not mandatory for sure, but it'd be nice to show it for the sake of completeness.

gsmet replied:
We decided against this one as with the Spring syntax, we would end up with `#_this`. I modified the Spring ScriptEvaluatorFactory to inject all the bindings anyway as it seems like a good thing to do.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/HibernateValidatorConfiguration.java:
```diff
@@ -62,6 +66,16 @@
+	 * {@link ScriptAssert} and {@link ParameterScriptAssert}
+	 * constraints. A fully qualified name of a class implementing  {@link ScriptEvaluatorFactory} is expected as a value.
+	 *
+	 * @since 6.1
```
On the question of the version to introduce this, my proposal is this:

* Let's add it in 6.0.3, marking all the members as `@Experimental` or `@Incubating`
* Promote to non-experimental in the future, e.g. 6.1 or a bit later, based on experiences
* Alter `@since` to 6.0.3 :)

I.e. let's bend the semantic versioning scheme a bit and take the liberty to introduce incubating features in a micro.

gsmet replied:
Done.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/internal/constraintvalidators/hv/AbstractScriptAssertValidator.java:
```diff
@@ -0,0 +1,52 @@
+	protected volatile ScriptAssertContext scriptAssertContext;
+
+	protected ScriptAssertContext getScriptAssertContext(ConstraintValidatorContext constraintValidatorContext) {
+		if ( scriptAssertContext == null ) {
```
Why is it done lazily upon `isValid()` and not via `initialize()`?

marko-bekhta replied:
That's actually that problem that @gsmet mentioned in another PR about having some sort of context in initialize method... As the factory is accessible only via `constraintValidatorContext` and there's no way of getting it during init, as we only have an annotation there, there's this lazy initialization ...

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/scripting/AbstractCachingScriptEvaluatorFactory.java:
```diff
@@ -0,0 +1,56 @@
+ * @author Kevin Pollet &lt;kevin.pollet@serli.com&gt; (C) 2011 SERLI
+ * @author Marko Bekhta
+ */
+public abstract class AbstractCachingScriptEvaluatorFactory implements ScriptEvaluatorFactory {
```
`@since` and `@Incubating` should be added.

gunnarmorling commented on engine/src/main/java/org/hibernate/validator/scripting/AbstractCachingScriptEvaluatorFactory.java:
```diff
@@ -0,0 +1,56 @@
+	/**
+	 * Retrieves a script executor for the given language.
+	 *
+	 * @param languageName The name of a scripting language as expected by {@link ScriptEngineManager#getEngineByName(String)}.
```
Is the reference to `ScriptEngineManager` correct here, as we abstract from JSR 223 now?

gunnarmorling commented on engine/src/test/java/org/hibernate/validator/test/internal/constraintvalidators/hv/ScriptAssertValidatorTest.java:
```diff
@@ -37,45 +38,55 @@
 	@Test
 	public void scriptEvaluatesToTrue() throws Exception {
-		ConstraintValidator<ScriptAssert, Object> validator = getInitializedValidator( "groovy", "true" );
+		@ScriptAssert(lang = "groovy", script = "true") class TmpType { }
```
Nice usage of local classes :)

gunnarmorling commented:
Ah, one more question: shouldn't this all go under `o.h.v.spi.scripting`? It's an extension point essentially which we mean to be implemented by advanced users.

marko-bekhta commented:
@gunnarmorling thanks for reviewing this once more :) if @gsmet don't mind - I can apply the suggested changes and rebase it to current master, as I see there are a couple of conflicts now. 
Also as I understand - this should go with 6.0.3, so I assume it should be ready for tomorrow's morning right?

gsmet commented:
@marko-bekhta don't worry, I'll do it :)

gunnarmorling commented:
> That's actually that problem that [...]

Ah, yes, got it. We should look into adding such extended contract indeed
then.

> so I assume it should be ready for tomorrow's morning right?

That'd be perfect, but please don't feel stressed about it. When it's done,
it's done, if not, then not :)

gunnarmorling commented:
>
> @marko-bekhta <https://github.com/marko-bekhta> don't worry, I'll do it :)
>

Ha, or that :) Thx!

gsmet commented:
And... merged! I rebased the PR and added 2 more commits to polish the documentation and JavaDoc.

@marko-bekhta thanks for your hard work!
